### PR TITLE
Feature/song meta data

### PR DESCRIPTION
**Pull request by teknopaul**
Hi long time since I've looked at LMMS code after github being weird with my fork.

But I have  had a go at making git happy with a simple change on one dedicated branch this time.  Not sure if anyone is interested, but this branch enables meta data for a song, i.e. title author and a image , that gets rendered into mp3s if you output as mp3.  Currently every mp3 make from LMMS requires use of a tag editor before publishing but this change means the song holds the mp3 meta data.

I think this meta data could be used for .ogg and .wav output too but I've not looked at that yet

### CODE REVIEW

**Comment by teknopaul** [comment]
in the auto checks, does complex method matter?

**Comment by zonkmachine** [comment]
> in the auto checks, does complex method matter?

I'm not sure, it didn't look that complex to me. I guess putting the `sf_set_string` stuff in a separate function would take care of it but maybe wait for a second opinion.

**Comment by teknopaul** [comment]
> > in the auto checks, does complex method matter?
> 
> I'm not sure, it didn't look that complex to me. I guess putting the `sf_set_string` stuff in a separate function would take care of it but maybe wait for a second opinion.

OK I'll do that now

**Comment by zonkmachine** [comment]
OK. Testing this and it seem to work mostly great. Some observations/questions.

* When I've added a picture, how do I remove it?
* Maybe there should be some default pictures to choose from?
* Also pictures could perhaps be added in an 'albumart' directory in ~/Documents/lmms
* When I add a picture it seem like there is need for a refresh of the Project Notes to account for the new size. Now it happens eventually but it should probably update when the picture is added.
projectnotesrefresh
projectotesrefresh2

File: src/core/Song.cpp [hunk 1]
```diff
@@ -1438,12 +1476,10 @@ void Song::setProjectFileName(QString const & projectFileName)
 	{
 		m_fileName = projectFileName;
 		emit projectFileNameChanged();
-	}
+    }
```
Comment (zonkmachine) [inline src/core/Song.cpp:1479]: Nope. We use tabs! Please change every instance of indents with spaces to tabs instead.

File: src/core/Song.cpp [hunk 2]
```diff
@@ -1243,6 +1271,16 @@ bool Song::saveProjectFile(const QString & filename, bool withResources)
 	m_masterVolumeModel.saveSettings( dataFile, dataFile.head(), "mastervol" );
 	m_masterPitchModel.saveSettings( dataFile, dataFile.head(), "masterpitch" );
 
+    QDomElement meta = dataFile.createElement(QString("meta"));
+    meta.setAttribute( "title", m_title );
```
Comment (zonkmachine) [inline src/core/Song.cpp:1275]: Spaces (the line above this is correct) ! 

```cpp
meta.setAttribute("title", m_title);
```

**Review by zonkmachine — COMMENTED** [review]
Some observations on coding style.
Also: https://github.com/LMMS/lmms/wiki/Coding-Conventions

**Comment by teknopaul** [comment]
Yeah sorry about whitespace, coding in the wrong tool. 
On it
What's opinion on public members vs getter setter? I come from Java which is opinionated.

**Comment by zonkmachine** [comment]
> Not sure if anyone is interested, but this branch enables meta data for a song, i.e. title author and a image ,

I believe I've seen more than one request for this. Nice! I've left you some thoughts and comments to get the review started. Please note that we have a monumental backlog of PR's since a code reorganization so forgive us if it takes a while.


> that gets rendered into mp3s if you output as mp3.

But this is not only for mp3 right? I've tested this on ogg and wav and the albumart is saved and shows up in my mediaplayer (Rhythmbox/Linux).

**Comment by teknopaul** [comment]
grrr branch seems to have lost changes to Song.cpp

**Comment by zonkmachine** [comment]
It compiles here.

Do `git diff` and `git status` to see what you've been up to lately.

**Comment by teknopaul** [comment]
I think trying to rebase master broke everything.  I have managed to commit whitespace fixes and added delete image to feature/song-meta-data, but presumably I still need to make this branch compatible with latest upstream master

> > Not sure if anyone is interested, but this branch enables meta data for a song, i.e. title author and a image ,
> 
> I believe I've seen more than one request for this. Nice! I've left you some thoughts and comments to get the review started. Please note that we have a monumental backlog of PR's since a code reorganization so forgive us if it takes a while.
> 
> > that gets rendered into mp3s if you output as mp3.
> 
> But this is not only for mp3 right? I've tested this on ogg and wav and the albumart is saved and shows up in my mediaplayer (Rhythmbox/Linux).

meta data is stored per song, but each format has different metadata rules. Only mp3 gets the image but song title and author should be there for ogg and wav.

**Comment by teknopaul** [comment]
if the wav writer API has ability to add images I didn't see it.

**Comment by teknopaul** [comment]
teknopaul@sbox:~/github_workspace/lmms$ git status
On branch feature/song-meta-data
Your branch is up to date with 'origin/master'.

& all changes gone :)

love git, gonna have to do the ol rm -rf and try again approach

**Comment by teknopaul** [comment]
hopefully whitespace fixed in this branch now

**Comment by teknopaul** [comment]
not sure about default image now, since it takes bytes in every save, was thinking could default username and date too but the current defaults are to not add the comment field at all so resulting file size is smaller.  Its not the case that these fields exist and can be populated or not, all 3 file formats support not having any meta data all.

**Comment by zonkmachine** [comment]
> not sure about default image now, since it takes bytes in every save, was thinking could default username and date too but the current defaults are to not add the comment field at all so resulting file size is smaller. Its not the case that these fields exist and can be populated or not, all 3 file formats support not having any meta data all.

Right, for default image, that's probably a no. Se the discussion in https://github.com/LMMS/lmms/issues/1245 which this PR would end up closing.

The new 'delete image' works like a charm.

File: src/core/audio/AudioFileWave.cpp [hunk 3]
```diff
@@ -88,7 +95,22 @@ bool AudioFileWave::startEncoding()
...
+	sf_set_string( m_sf, SF_STR_SOFTWARE, "LMMS" );
+
+	const Song* song = Engine::getSong();
+	if ( !song->getTitle().isEmpty() )
+	{
+		addComment(SF_STR_TITLE, song->getTitle());
+	} else {
+		addComment(SF_STR_TITLE, QFileInfo(song->projectFileName())
+												.completeBaseName()
+												.replace("[_-]", " "));
```
[the remark this answers is not in the document]
Comment (teknopaul) [reply src/core/audio/AudioFileWave.cpp:107]: afaik im using 4 but github might nit be

**Comment by teknopaul** [comment]
I added more fields for flac and support for images in ogg files.  Should be possible to add images for flac and wav but we use libsndfile which has a missing feature  https://github.com/libsndfile/libsndfile/issues/982

File: src/core/audio/AudioFileWave.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zonkmachine) [reply src/core/audio/AudioFileWave.cpp:107]: Yes it's github. They apparently default to 8. You can add `?ts=4` to the url to set the tab width to 4.

**Comment by teknopaul** [comment]
pushed a fix re: running out of buffer when adding big images

**Comment by DomClark** [comment]
Could you group all the metadata together into one `Metadata` class? It's all used together, saved together in its own element, and `Song` has too much in it already.

**Comment by tresf** [comment]
@teknopaul first, thanks for this contribution, I agree that using a 3rd party tool to edit the MP3 metadata is cumbersome.

I notice that you've closed this pull request, but i believe that there's value in keeping it around.  Hopefully the multiple PRs, issues and feedback thereof are not discouraging.  We're happy to see these contributions. :)

**Pull request closed without merging**